File: src/tools/screenshot.ts
import { z } from "zod";
import { TerminalManager, SessionManager } from "../terminal/index.js";

export const screenshotSchema = z.object({});

export const screenshotSchemaWithSession = z.object({
  sessionId: z.string().optional().describe("Session ID (optional, uses default session if not provided)"),
});

export type ScreenshotArgs = z.infer<typeof screenshotSchema>;
export type ScreenshotArgsWithSession = z.infer<typeof screenshotSchemaWithSession>;

export const screenshotTool = {
  name: "takeScreenshot",
  description: "Capture terminal state as structured JSON with: content (visible text), cursor {x, y} position, and dimensions {cols, rows}. Use when you need cursor position (e.g., for interactive apps, vim) or terminal dimensions. For simple command output, prefer getContent().",
  inputSchema: {
    type: "object" as const,
    properties: {},
    required: [],
  },
};

export const screenshotToolWithSession = {
  name: "takeScreenshot",
  description: "Capture terminal state as structured JSON with: content (visible text), cursor {x, y} position, and dimensions {cols, rows}. Use when you need cursor position (e.g., for interactive apps, vim) or terminal dimensions. For simple command output, prefer getContent().",
  inputSchema: {
    type: "object" as const,
    properties: {
      sessionId: {
        type: "string",
        description: "Session ID (optional, uses default session if not provided)",
      },
    },
    required: [],
  },
};

export function handleScreenshot(manager: TerminalManager, _args: unknown): { content: Array<{ type: "text"; text: string }> } {
  const screenshot = manager.takeScreenshot();

  const result = {
    content: screenshot.content,
    cursor: screenshot.cursor,
    dimensions: screenshot.dimensions,
  };

  return {
    content: [
      {
        type: "text",
        text: JSON.stringify(result, null, 2),
      },
    ],
  };
}

export async function handleScreenshotWithSession(
  manager: SessionManager,
  args: unknown
): Promise<{ content: Array<{ type: "text"; text: string }> }> {
  const parsed = screenshotSchemaWithSession.parse(args);
  const session = await manager.getSession(parsed.sessionId);
  const screenshot = session.takeScreenshot();

  const result = {
    content: screenshot.content,
    cursor: screenshot.cursor,
    dimensions: screenshot.dimensions,
  };

  return {
    content: [
      {
        type: "text",
        text: JSON.stringify(result, null, 2),
      },
    ],
  };
}

File: src/tools/screenshot.rs
use serde::Deserialize;
use serde_json::Value;

use crate::mcp::{CallToolResult, ToolDefinition};
use crate::terminal::Terminal;

/// Arguments for the takeScreenshot tool
#[derive(Debug, Deserialize, Default)]
pub struct ScreenshotArgs {
    /// Format of the screenshot: "ansi" (default) or "plain"
    #[serde(default)]
    pub format: Option<String>,
}

/// Take a screenshot of the terminal, optionally with ANSI colors
pub fn take_screenshot(terminal: &mut Terminal, args: Value) -> Result<CallToolResult, String> {
    let args: ScreenshotArgs = serde_json::from_value(args).unwrap_or_default();

    let format = args.format.as_deref().unwrap_or("ansi");

    let content = match format {
        "plain" => terminal.get_content(),
        "ansi" | _ => terminal.take_screenshot(),
    };

    // Get cursor position for additional context
    let (row, col) = terminal.cursor_position();
    let (cols, rows) = terminal.size();

    // Build a header with terminal info
    let header = format!(
        "Terminal: {}x{} | Cursor: ({}, {})\n{}\n",
        cols, rows, row, col,
        "─".repeat(cols as usize)
    );

    Ok(CallToolResult::text(format!("{}{}", header, content)))
}

/// Get the tool definition for takeScreenshot
pub fn definition() -> ToolDefinition {
    ToolDefinition {
        name: "takeScreenshot".to_string(),
        description: "Capture the current terminal state as text. By default includes ANSI escape codes for colors and formatting. Use format='plain' for plain text without escape codes.".to_string(),
        input_schema: serde_json::json!({
            "type": "object",
            "properties": {
                "format": {
                    "type": "string",
                    "description": "Output format: 'ansi' (default) includes color codes, 'plain' is plain text only",
                    "enum": ["ansi", "plain"],
                    "default": "ansi"
                }
            }
        }),
    }
}
